Add an optional query parameter to SSO redirect (#3824)
matrix-org/matrix-spec-proposals: proposals/3824-oidc-aware-clients.md

Secondly, the homeserver can optionally specify which auth type is `delegated_oidc_compatibility` are supported for an authentication type.

### Homeserver indicates that an `m.login.sso` flow is for compatibility

Add an optional `delegated_oidc_compatibility` field to the response of `GET /_matrix/client/v3/login`:

`"delegated_oidc_compatibility"?: boolean`

For example, if a homeserver is advertising password login for legacy clients only then it could return the following:

```
{
  "flows": [{
    "type": "m.login.password"
  }, {
    "type": "m.login.sso",
    "delegated_oidc_compatibility": true
  }]
}

```

If the client finds `delegated_oidc_compatibility` to be `true` then, assuming it supports that auth type, it should present this as the only login/registration method available to the user.

### Client indicates `action` on SSO redirect

Add an optional query parameter `action` to `GET /_matrix/client/v3/login/sso/redirect` and `GET /_matrix/client/v3/login/sso/redirect/{idpId}` with meaning:

- `login` - the SSO redirect is for the purposes of signing an existing user in

> - `register` - the SSO redirect is for the purpose of registering a new user account

Reviewer: I think I'm missing something -- how would the client know whether it is logging in or registering?

Author: The user would have told the client if they already have a Matrix account or not.

For example, by clicking buttons labelled as such.

e.g. `https://matrix-client.matrix.org/_matrix/client/v3/login/sso/redirect?action=register`

n.b. we don't need to add this to the [Login Fallback](https://spec.matrix.org/v1.2/client-server-api/#login-fallback) as that isn't used for registration.

### Definition of OIDC aware

For a client to be considered *OIDC aware* it would:

- support the `m.login.sso` auth flow
- where a `delegated_oidc_compatibility` value of `true` is present on an `m.login.sso` then *only* offer that auth flow to the user
- append `action=login` and `action=register` parameters to the SSO redirect URLs

> - sign post and link users to manage their account at the OP web UI given by [MSC2965](https://github.com/matrix-org/matrix-spec-proposals/pull/2965)

Reviewer: I originally took "sign" to be crypto-graphically sign.

```suggestion
- Link users to manage their account at the OP web UI given by [MSC2965](https://github.com/matrix-org/matrix-spec-proposals/pull/2965) instead of native UI
```

- RECOMMENDED: label the SSO button as "Continue"

For an OIDC enabled homeserver to provide support for *OIDC aware* clients it would:

- support OIDC delegation as per [MSC2964](https://github.com/matrix-org/matrix-spec-proposals/pull/2964) and others

> - provide a compatibility layer for `m.login.password` and `m.login.sso` that wraps on to OIDC

Reviewer: Is it defined how `m.login.password` an be wrapped into OIDC? I guess doing the login itself?

  }]
}

```

If the client finds `delegated_oidc_compatibility` to be `true` then, assuming it supports that auth type, it should present this as the only login/registration method available to the user.

### Client indicates `action` on SSO redirect

Add an optional query parameter `action` to `GET /_matrix/client/v3/login/sso/redirect` and `GET /_matrix/client/v3/login/sso/redirect/{idpId}` with meaning:

- `login` - the SSO redirect is for the purposes of signing an existing user in
- `register` - the SSO redirect is for the purpose of registering a new user account

e.g. `https://matrix-client.matrix.org/_matrix/client/v3/login/sso/redirect?action=register`

n.b. we don't need to add this to the [Login Fallback](https://spec.matrix.org/v1.2/client-server-api/#login-fallback) as that isn't used for registration.

### Definition of OIDC aware

For a client to be considered fully *OIDC aware* it **must**:

- support the `m.login.sso` auth flow
- where a `delegated_oidc_compatibility` value of `true` is present on an `m.login.sso` then *only* offer that auth flow to the user
- append `action=login` and `action=register` parameters to the SSO redirect URLs

> - link users to manage their account at the OpenID Provider web UI given by [MSC2965](https://github.com/matrix-org/matrix-spec-proposals/pull/2965) instead of native UI

Reviewer 2: I am guessing that this and other references to MSC2965 here refer to the `account_management_uri` which has moved to MSC4191?

Author: Yes, done in [`4188601` (#3824)](https://github.com/matrix-org/matrix-spec-proposals/pull/3824/commits/4188601687630fda218af4e811953d8953ad17a3)

- check and honour the `m.3pid_changes` [capability](https://spec.matrix.org/v1.7/client-server-api/#m3pid_changes-capability) so that the user is not offered the ability to add or remove 3PIDs when OIDC is used
- if the user wishes to sign out a device session other than it's own then the client **must**:
  - link the user to the OpenID Provider web UI given by [MSC2965](https://github.com/matrix-org/matrix-spec-proposals/pull/2965) if provided
  - append the `action` and `device_id` to the web UI link parameters described by [MSC2965](https://github.com/sandhose/matrix-doc/blob/msc/sandhose/oidc-discovery/proposals/2965-oidc-discovery.md#account-management-url-parameters) so that the web UI knows that the user wishes to sign out a device and which one it is. e.g. `?action=session_end&device_id=<device_id>`

Optionally, an *OIDC aware* client **could**:

> - label the SSO button as "Continue"

Author: I have elaborated on this in the latest version.

# MSC3824: OAuth 2.0 API aware clients

As of spec 1.15 we now have two APIs for clients to authenticate:

- the [Legacy API](https://spec.matrix.org/v1.15/client-server-api/#legacy-api)
- the [OAuth 2.0 API](https://spec.matrix.org/v1.15/client-server-api/#oauth-20-api)

In this context we can define four types of client:

1. *OAuth native client* - This is a client that, where the homeserver supports it, uses the OAuth 2.0 API for login and registration. e.g. Element X, Element Web

> 1. *OAuth aware client* - This is a client that is aware of the OAuth 2.0 API but will still uses the Legacy PI (e.g. `m.login.sso`) to auth with an OAuth 2.0 API enabled homeserver.

Reviewer 3: ```suggestion
1. *OAuth aware client* - This is a client that is aware of the OAuth 2.0 API but will still uses the Legacy API (e.g. `m.login.sso`) to auth with an OAuth 2.0 API enabled homeserver.
```
?

> 1. *OAuth aware client* - This is a client that is aware of the OAuth 2.0 API but will still uses the Legacy PI (e.g. `m.login.sso`) to auth with an OAuth 2.0 API enabled homeserver.

Reviewer 3: Oh never mind, but it could use some quotes around 'aware' and a 'see below'.

> 1. *OAuth aware client* - This is a client that is aware of the OAuth 2.0 API but will still uses the Legacy PI (e.g. `m.login.sso`) to auth with an OAuth 2.0 API enabled homeserver.

Author: Included in efc0af9e356075c227cf1f683c73aefac3cf3018

1. *Legacy client with SSO support* - This is a client that is not aware of the OAuth 2.0 API but does support the `m.login.sso` flow. e.g. Element Classic on iOS and Android
1. *Legacy client without SSO support* - This is a client that is not aware of the OAuth 2.0 API at all and nor does it support `m.login.sso` flow. Typically auth is done via `m.login.password` only. e.g. Fractal

The purpose of differentiating #2 and #3 is that, for a Legacy client with SSO support, the user journey can be optimised with minimal modifications when talking to an OAuth 2.0 API enabled homeserver.

This proposal outlines changes to facilitate clients in becoming OAuth 2.0 API aware.

## Proposal

> Firstly, a client can specify which action the user is wanting to achieve at the point of SSO redirection. This allows the homeserver to display the most relevant UI to the user.

Author: Fair point. I've fixed this in 38cdbd81012a1e478f67130680cc344004bc481a

Secondly, the homeserver can optionally specify which auth type is `delegated_oidc_compatibility` are supported for an authentication type.

### Homeserver indicates that an `m.login.sso` flow is for compatibility

Add an optional `delegated_oidc_compatibility` field to the response of `GET /_matrix/client/v3/login`:

`"delegated_oidc_compatibility"?: boolean`

For example, if a homeserver is advertising password login for legacy clients only then it could return the following:

```json
{
  "flows": [{
    "type": "m.login.password"
  }, {
    "type": "m.login.sso",
    "delegated_oidc_compatibility": true
  }]
}

```

> If the client finds `delegated_oidc_compatibility` to be `true` then, assuming it supports that auth type, it should present this as the only login/registration method available to the user.

Reviewer 3: Sorry, I'm not getting this at all. This is intended for legacy clients? Both type 3 and type 4 in the table above? Does this mean legacy clients do have to make some change? The intention is that the HS takes the username/password and forwards it to the oauth provider as a legacy compatibility measure? Is there a specific scenario in which clients like this are erroneously presenting another flow?
